lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1247,24 +1254,25 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
 
   setPrivateDiscriminatorIfNeeded(IRGenOpts, MSF);
 
-  (void)printAsObjCIfNeeded(
-      opts.InputsAndOutputs.supplementaryOutputs().ObjCHeaderOutputPath,
-      Instance.getMainModule(), opts.ImplicitObjCHeaderPath, moduleIsPublic);
+  (void)printAsObjCIfNeeded(PSPs.SupplementaryOutputs.ObjCHeaderOutputPath,
+                            Instance.getMainModule(),
+                            opts.ImplicitObjCHeaderPath, moduleIsPublic);
 
   if (Action == FrontendOptions::ActionType::EmitSIB)
-    return serializeSIB(Invocation.getFrontendOptions(), SM.get(),
+    return serializeSIB(Invocation.getFrontendOptions(), SM.get(), PSPs,
                         Instance.getASTContext(), MSF);
 
   const bool haveModulePath =
-      !opts.InputsAndOutputs.supplementaryOutputs().ModuleOutputPath.empty() ||
-      !opts.InputsAndOutputs.supplementaryOutputs().ModuleDocOutputPath.empty();
+      !PSPs.SupplementaryOutputs.ModuleOutputPath.empty() ||
+      !PSPs.SupplementaryOutputs.ModuleDocOutputPath.empty();
   if (haveModulePath && !SM->isSerialized())
     SM->serialize();
 
   if (haveModulePath) {
     if (Action == FrontendOptions::ActionType::MergeModules ||
         Action == FrontendOptions::ActionType::EmitModuleOnly) {
-      return emitIndexDataIfNeeded(Instance.getPrimarySourceFile(), Invocation,
+      // FIXME: what if MSF is a module?
+      return emitIndexDataIfNeeded(MSF.dyn_cast<SourceFile *>(), Invocation,
```

Comment:
`emitIndexDataIfNeeded` already handles that case; it'll index everything. You could change it to take a ModuleOrSourceFile if it's bothering you. (It's currently getting the module out of the CompilerInstance, which isn't really the best.)

Reply:
Changed the FIXME to a comment including your answer.